File: missing-statement-descriptor/node/stripe-statement-descriptor.mjs
/**
 * Report Stripe accounts whose statement descriptor is missing or inconsistent.
 *
 * Read only. Three paginated GETs and no writes: give this a RESTRICTED key with
 * read access to Account, Charges and Disputes. The repair is printed, never run.
 */
const API = 'https://api.stripe.com/v1';

export const MIN_LEN = 5;
export const MAX_LEN = 22;
const MIN_LETTERS = 5;
const BANNED = ['<', '>', "'", '"'];
const UNRECOGNISED = new Set(['unrecognized', 'general', 'duplicate']);

/**
 * Classify the account's descriptor. Pure, so the format rules test offline.
 * `descriptors` is every calculated_statement_descriptor seen on recent charges.
 */
export function verdict(prefix, descriptors) {
  const seen = [...new Set((descriptors ?? [])
    .map((d) => (d ?? '').trim())
    .filter((d) => d !== ''))].sort();

  if (!(prefix ?? '').trim()) {
    return ['unset',
      `no statement descriptor prefix on the account; ${seen.length} distinct ` +
      'descriptor(s) observed on charges'];
  }
  if (descriptors && descriptors.length && seen.length === 0) {
    return ['blank',
      'a prefix is configured but every charge carried an empty descriptor: ' +
      'nothing identifying you reaches the networks'];
  }
  if (seen.length > 1) {
    return ['fragmented',
      `${seen.length} distinct descriptors in use (${seen.slice(0, 3).join(', ')}): ` +
      'Visa identifies a monitored account by the static component, so your ' +
      'volume is being split'];
  }
  const text = seen.length ? seen[0] : prefix.trim();
  const letters = [...text].filter((c) => /[a-z]/i.test(c)).length;
  if (text.length < MIN_LEN || text.length > MAX_LEN) {
    return ['malformed',
      `"${text}" is ${text.length} characters; Stripe requires ${MIN_LEN} to ${MAX_LEN}`];
  }
  if (letters < MIN_LETTERS) {
    return ['malformed',
      `"${text}" has ${letters} letter(s); Stripe requires at least ${MIN_LETTERS}`];
  }
  if (BANNED.some((c) => text.includes(c))) {
    return ['malformed',
      `"${text}" contains a character Stripe disallows in a descriptor`];
  }
  return ['consistent', `${text} across the sampled charges`];
}

async function get(key, path, params = {}) {
  const url = new URL(API + path);
  for (const [k, v] of Object.entries(params)) url.searchParams.set(k, v);
  const res = await fetch(url, { headers: { Authorization: `Bearer ${key}` } });
  if (res.status === 401) {
    throw new Error('401 from Stripe: the key is wrong, or is for the other mode');
  }
  if (!res.ok) throw new Error(`${res.status} from ${url.pathname}`);
  return res.json();
}

async function page(key, path, cap, params = {}) {
  const out = [];
  const q = { ...params, limit: 100 };
  for (;;) {
    const p = await get(key, path, q);
    const data = p.data ?? [];
    out.push(...data);
    if (data.length === 0 || !p.has_more || out.length >= cap) break;
    q.starting_after = data[data.length - 1].id;
  }
  return out;
}

async function main() {
  const key = (process.env.STRIPE_API_KEY || "dummy-stripe-api-key");
  if (!key) {
    console.error('set STRIPE_API_KEY (use a restricted, read-only key)');
    process.exitCode = 2;
    return;
  }

  const account = await get(key, '/account');
  const settings = account.settings ?? {};
  const prefix = settings.card_payments?.statement_descriptor_prefix
    ?? settings.payments?.statement_descriptor;

  const days = Number((process.env.DAYS || "dummy-days") ?? 30);
  const disputeDays = Number((process.env.DISPUTE_DAYS || "dummy-dispute-days") ?? 180);
  const now = Date.now() / 1000;
  const charges = await page(key, '/charges', 5000,
    { 'created[gte]': Math.floor(now - days * 86400) });
  const descriptors = charges.map((c) => c.calculated_statement_descriptor);
  const noSuffix = charges.filter((c) => !c.statement_descriptor_suffix).length;

  const [state, detail] = verdict(prefix, descriptors);
  console.log(`${state.padEnd(11)} ${detail} (${charges.length} charge(s) sampled)`);

  const disputes = await page(key, '/disputes', 1000,
    { 'created[gte]': Math.floor(now - disputeDays * 86400) });
  if (disputes.length) {
    const blind = disputes.filter((d) => UNRECOGNISED.has(d.reason)).length;
    console.log(`${(100 * blind / disputes.length).toFixed(1)}% of disputes cite ` +
                `unrecognized, general or duplicate (${blind} of ${disputes.length})`);
  }

  if (state === 'consistent' && !noSuffix) return;

  if (state !== 'consistent') {
    console.warn('  set the prefix in Dashboard, Settings, Business, public business ' +
                 `information: ${MIN_LEN} to ${MAX_LEN} characters, at least ` +
                 `${MIN_LETTERS} letters, and none of ${BANNED.join(' ')}`);
    console.warn('  use the website domain or the business name customers know, and ' +
                 'keep it identical across every payment flow');
  }
  if (noSuffix) {
    console.warn(`  ${noSuffix} of ${charges.length} charge(s) carried no ` +
                 'statement_descriptor_suffix. Set one at payment creation so the ' +
                 'line names the order.');
  }
  process.exitCode = 1;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing key, and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}

File: missing-statement-descriptor/python/stripe_statement_descriptor.py
"""Report Stripe accounts whose statement descriptor is missing or inconsistent.

Read only. Three paginated GETs and no writes: give this a RESTRICTED key with
read access to Account, Charges and Disputes. The repair is printed, never
performed, because this script holds a credential to a live payments account.
"""
import argparse
import logging
import os
import sys
import time

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("stripe_statement_descriptor")

API = "https://api.stripe.com/v1"

MIN_LEN = 5
MAX_LEN = 22
MIN_LETTERS = 5
BANNED = ("<", ">", "'", '"')

# Reason codes a customer gives when the line on the statement meant nothing.
UNRECOGNISED = ("unrecognized", "general", "duplicate")


def verdict(prefix, descriptors):
    """Classify the account's descriptor. Pure, so the format rules test offline.

    `prefix` is the configured static prefix; `descriptors` is every
    calculated_statement_descriptor observed on recent charges. Returns
    (state, detail).
    """
    seen = sorted({(d or "").strip() for d in (descriptors or [])} - {""})
    if not (prefix or "").strip():
        return ("unset",
                "no statement descriptor prefix on the account; %d distinct "
                "descriptor(s) observed on charges" % len(seen))
    if descriptors and not seen:
        return ("blank",
                "a prefix is configured but every charge carried an empty "
                "descriptor: nothing identifying you reaches the networks")
    if len(seen) > 1:
        return ("fragmented",
                "%d distinct descriptors in use (%s): Visa identifies a monitored "
                "account by the static component, so your volume is being split"
                % (len(seen), ", ".join(seen[:3])))
    text = seen[0] if seen else prefix.strip()
    letters = sum(1 for c in text if c.isalpha())
    if len(text) < MIN_LEN or len(text) > MAX_LEN:
        return ("malformed",
                "%r is %d characters; Stripe requires %d to %d"
                % (text, len(text), MIN_LEN, MAX_LEN))
    if letters < MIN_LETTERS:
        return ("malformed",
                "%r has %d letter(s); Stripe requires at least %d"
                % (text, letters, MIN_LETTERS))
    if any(c in text for c in BANNED):
        return ("malformed",
                "%r contains a character Stripe disallows in a descriptor" % text)
    return ("consistent", "%s across the sampled charges" % text)


def get(session, path, **params):
    r = session.get(API + path, params=params, timeout=30)
    if r.status_code == 401:
        raise SystemExit("401 from Stripe: the key is wrong, or is for the other mode")
    r.raise_for_status()
    return r.json()


def page(session, path, cap, **params):
    out = []
    params = dict(params)
    params["limit"] = 100
    while True:
        p = get(session, path, **params)
        data = p.get("data", [])
        out.extend(data)
        if not data or not p.get("has_more") or len(out) >= cap:
            break
        params["starting_after"] = data[-1]["id"]
    return out


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--days", type=int, default=30,
                    help="how far back to sample charges")
    ap.add_argument("--dispute-days", type=int, default=180,
                    help="how far back to read disputes")
    ap.add_argument("--max-charges", type=int, default=5000,
                    help="stop paginating after this many charges")
    args = ap.parse_args()

    key = os.environ.get("STRIPE_API_KEY")
    if not key:
        log.error("set STRIPE_API_KEY (use a restricted, read-only key)")
        return 2

    s = requests.Session()
    s.headers.update({"Authorization": "Bearer " + key})

    account = get(s, "/account")
    settings = account.get("settings") or {}
    prefix = ((settings.get("card_payments") or {}).get("statement_descriptor_prefix")
              or (settings.get("payments") or {}).get("statement_descriptor"))

    now = time.time()
    charges = page(s, "/charges", args.max_charges,
                   **{"created[gte]": int(now - args.days * 86400)})
    descriptors = [c.get("calculated_statement_descriptor") for c in charges]
    no_suffix = sum(1 for c in charges if not c.get("statement_descriptor_suffix"))

    state, detail = verdict(prefix, descriptors)
    log.info("%-11s %s (%d charge(s) sampled)", state, detail, len(charges))

    disputes = page(s, "/disputes", 1000,
                    **{"created[gte]": int(now - args.dispute_days * 86400)})
    if disputes:
        blind = sum(1 for d in disputes if d.get("reason") in UNRECOGNISED)
        log.info("%.1f%% of disputes cite unrecognized, general or duplicate (%d of %d)",
                 100.0 * blind / len(disputes), blind, len(disputes))

    if state == "consistent" and not no_suffix:
        return 0

    if state != "consistent":
        log.warning("  set the prefix in Dashboard, Settings, Business, public "
                    "business information: %d to %d characters, at least %d letters, "
                    "and none of %s", MIN_LEN, MAX_LEN, MIN_LETTERS, " ".join(BANNED))
        log.warning("  use the website domain or the business name customers know, "
                    "and keep it identical across every payment flow")
    if no_suffix:
        log.warning("  %d of %d charge(s) carried no statement_descriptor_suffix. Set "
                    "one at payment creation so the line names the order.",
                    no_suffix, len(charges))
    return 1


if __name__ == "__main__":
    sys.exit(main())
